### DISCUSSION TITLE
href always prefixed even when `localePrefix` is `never`

### DISCUSSION THREAD

**Discussion by y-nk** [amannn/next-intl#1328] [Q&A] [answered]
### Description

I'm creating a website and want to localize urls directly, therefore i dont need locale prefixing. i'm using the `localePrefix: 'never'` option but i can't get to generate internal links for another locale within the page because there is an added condition which forces prefixing if the locale of the link is not the current locale:

https://github.com/amannn/next-intl/blob/241c3ea2b0c7c8444b76518e0a3a2979aa208497/packages/next-intl/src/navigation/shared/BaseLink.tsx#L34-L49

with above code, having a `localePrefix: 'never'` but trying to generate a `<Link locale="fr" />` when the current locale is `en` will result in the link href to be prefixed with `/fr/`, which defeats the `localePrefix: 'never'` strategy.

### Verifications

- [X] I've verified that the problem I'm experiencing isn't covered in [the docs](https://next-intl-docs.vercel.app/).
- [X] I've searched for similar, existing issues on GitHub and [Stack Overflow](https://stackoverflow.com/search?q=next-intl).
- [X] I've compared my app to [a working example](https://next-intl-docs.vercel.app/examples) to look for differences.

### Mandatory reproduction URL

https://example.com

### Reproduction description

I'm still working on a reproduction url, but i believe `isChangingLocale` should be removed from the code above.

### Expected behaviour

When using `localPrefix: 'never'`, urls should never be prefixed with locale.

**Comment by amannn** [comment]
The reason why links to other locales have a locale prefix even though you're using `localePrefix: 'never'` is because this route is used to update your `NEXT_LOCALE` cookie. The cookie is the source of truth in this mode and therefore it's important to update it first before the final route you're intending to navigate to can be visited. You should see a redirect from `/fr` back to `/` once the link is clicked.

Does this affect your app in any negative way or is this just something you've noticed and it looked like a bug?

**Comment by y-nk** [comment]
Yes, it has a slightly negative impact because the goal is to have urls matching exactly the urls i set for SEO purpose not the ones with the prefixes, especially that i do not have those prefixes in the `pathnames` that i generate which i also use for my sitemap (so i have a sitemap with `/hello` but the links in my page are `/en/hello`.

I still try to make a reproduction repo but it's a bit entangled with other libraries therefore it takes time to isolate.

whould that behavior be disabled when i use `localeDetection: false` ?

**Comment by amannn** [comment]
If SEO is important to you, how do you provide distinct pathnames per locale if you're using `localePrefix: 'never'`? Do you use domain-based routing? It would be helpful if you could share your routing config.

> whould that behavior be disabled when i use localeDetection: false ?

Nope, that doesn't change this behavior.

**Comment by y-nk** [comment]
i'm managing urls manually - the website is built with next and [content-collections](https://content-collections.dev) so i'm feeding slugs manually (similar to subdomain-based routing with a single domain)

**Comment by amannn** [comment, marked as answer]
If you're using a routing system that is external to `next-intl`, maybe you'd have an easier time if you'd set up your app [without i18n routing](https://next-intl-docs.vercel.app/docs/getting-started/app-router). Using that approach, you don't have to use either the middleware or the navigation APIs from `next-intl` and also not the `[locale]` segment. Maybe that puts you in a better place where you can implement your custom routing requirements.

I'm going to move this to a discussion since this seems to be more a usage question than a bug report. As explained above, the prefix that is shown when changing a locale serves a concrete purpose and is working as intended.
